client/comms/wsconn.go

```diff
@@ -0,0 +1,377 @@
+package comms
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/decred/dcrdex/server/comms/msgjson"
+	"github.com/gorilla/websocket"
+)
+
+const (
+	// pingWait is the maximum time in seconds to wait for a ping from
+	// the server. The client can be reconnected after this threshold since
+	// the websocket connection to the server is assumed to broken.
+	pingWait = 60 * time.Second
+
+	// writeWait is the maximum time in seconds to write to the connection.
+	writeWait = 5 * time.Second
+
+	// bufferSize is buffer size for the websocket connection's read channel.
+	readBuffSize = 128
+)
+
+// WsCfg is the configuration struct for initializing a WsConn.
+type WsCfg struct {
+	// The websocket host.
+	Host string
+	// The websocket api path.
+	Path string
+	// The maximum time in seconds to wait for a ping from the server.
+	PingWait time.Duration
+	// The maximum time in seconds to write to the connection.
+	WriteWait time.Duration
+	// The rpc certificate file path.
+	RpcCert string
+	// The rpc key file path.
+	RpcKey string
+	// Skip server cert verification, should only be used in testing.
+	InsecureSkipVerify bool
+	// ReconnectSync runs the needed reconnection synchronisation after
+	// a disconnect.
+	ReconnectSync func()
+	// The context.
+	Ctx context.Context
+	// The context cancel func.
+	Cancel context.CancelFunc
+}
+
+// WsConn represents a client websocket connection.
+type WsConn struct {
+	rID          uint64
+	cfg          *WsCfg
+	ws           *websocket.Conn
+	tlsCfg       *tls.Config
+	readCh       chan *msgjson.Message
+	sendCh       chan *msgjson.Message
+	req          map[uint64]*msgjson.Message
+	reqMtx       sync.RWMutex
+	writeMtx     sync.Mutex
+	connected    bool
+	connectedMtx sync.RWMutex
+	once         sync.Once
+	wg           sync.WaitGroup
+}
+
+// filesExists reports whether the named file or directory exists.
+func fileExists(name string) bool {
+	_, err := os.Stat(name)
+	return !os.IsNotExist(err)
+}
+
+// NewWsConn creates a client websocket connection.
+func NewWsConn(cfg *WsCfg) (*WsConn, error) {
+	if cfg.PingWait < time.Second {
+		return nil, fmt.Errorf("ping wait time should be a second or greater")
+	}
+
+	if cfg.WriteWait < time.Second {
+		return nil, fmt.Errorf("write wait time should be a second or greater")
+	}
+
+	var tlsConfig *tls.Config
+	if fileExists(cfg.RpcCert) && fileExists(cfg.RpcKey) {
+		keypair, err := tls.LoadX509KeyPair(cfg.RpcCert, cfg.RpcKey)
+		if err != nil {
+			return nil, err
+		}
+
+		// Prepare the TLS configuration.
+		tlsConfig = &tls.Config{
+			Certificates: []tls.Certificate{keypair},
+			MinVersion:   tls.VersionTLS12,
+		}
+
+		if cfg.InsecureSkipVerify {
+			tlsConfig.InsecureSkipVerify = true
+		}
```

Comment:
I'd say that if a certificate is provided, we should assume it's self-signed and set `InsecureSkipVerify = true`. You can drop the field in the `WsCfg` struct if we do it that way.

Reply:
I'm not sure about this.  My understanding of `InsecureSkipVerify` is that when true, it skips validation of the server certificate.  This only makes sense in my mind if we are not able to specify the correct server certificate (in `RpcCert`) and you want to take your chances that there's no man-in-the-middle attack.  When we provide the server certificate to the client config, we are then able to validate the certificate chain and hostname from the server, so I think we do want this false in production.